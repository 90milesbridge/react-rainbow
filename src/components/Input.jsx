import React from 'react';
import classnames from 'classnames';
import InputError from './InputError.jsx';
import InputFile from './InputFile.jsx';
import uniqueId from './libs/uniqueId';

let isRegularInput = (type) => ['text', 'search', 'number', 'email', 'password'].indexOf(type) !== -1;
let isTypeFile = (type) => type === 'file';

export default class Input extends React.Component {

    componentWillMount() {
        this.inputId = uniqueId('input');
    }

    render() {
        if (isRegularInput(this.props.type)) return this.renderRegularInput();
        if (isTypeFile) return <InputFile id={ this.inputId } {...this.props}/>
    }

    renderRegularInput() {
        let inputContainerClasses = classnames('slds-form-element', {
            'slds-has-error' : !this.props.isValid }, this.props.className);

        return (
            <div className={ inputContainerClasses }>
                <label className="slds-form-element__label" htmlFor={ this.inputId }>
                    { this.props.required ? <abbr className="slds-required" title="required">*</abbr> : null }
                    { this.props.label }
                </label>
                <div className="slds-form-element__control">
                    <input id={ this.inputId }
                           ref={ (input) => this.input = input }
                           className="slds-input"
                           onInput={ (e) => this.handleOnInput(e) }
                           onBlur={ (e) => this.handleOnBlur(e) }
                           onFocus={ (e) => this.handleOnFocus(e) }
                           type={ this.props.type }
                           step={ this.props.step }
                           placeholder={ this.props.placeholder }
                           required={ this.props.required }
                           disabled={ this.props.disabled }
                           name={ this.props.name }
                    />
                </div>
                <InputError isVisible={ !this.props.isValid } message="The field is invalid."/>
            </div>
        );
    }

    handleOnInput(e) {
        this.props.onChange && this.props.onChange(e);
    }

    handleOnBlur(e) {
        this.props.onBlur && this.props.onBlur(e);
    }

    handleOnFocus(e) {
        this.props.onFocus && this.props.onFocus(e);
    }

}

Input.propTypes = {
    label: React.PropTypes.string.isRequired,
    placeholder: React.PropTypes.string,
    required: React.PropTypes.bool,
    disabled: React.PropTypes.bool,
    step: React.PropTypes.number,
    name: React.PropTypes.string,
    isValid: React.PropTypes.bool.isRequired,
    onChange: React.PropTypes.func
};

Input.defaultProps = {
    type: 'text',
    required: false,
    disabled: false,
    step: 0.01
};