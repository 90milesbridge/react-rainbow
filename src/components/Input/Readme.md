input base

    <div style={{padding: '1rem'}}>
        <Input placeholder="Placeholder text" iconName="utility:activity" />
    </div>

input error
    <div style={{padding: '1rem'}}>
        <Input placeholder="Placeholder text" error="Error" iconName="utility:activity" label="Input label" required />
    </div>

input help-icon
    <div style={{padding: '1rem'}}>
        <Input placeholder="Placeholder text" bottomHelpText="Help text" iconName="utility:activity" label="Input label" />
    </div>
