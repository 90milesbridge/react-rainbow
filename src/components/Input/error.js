import React from 'react';
import PropTypes from 'prop-types';

export default function Error({ error }) {
    if (error) {
        return <div className="slds-form-element__help">{error}</div>;
    }
    return null;
}

Error.propTypes = {
    error: PropTypes.node.isRequired,
};
