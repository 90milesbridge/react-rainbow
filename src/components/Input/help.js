import React from 'react';
import PropTypes from 'prop-types';

export default function Help({ text }) {
    if (text) {
        return <div id="123" className="slds-form-element__help">{text}</div>;
    }
    return null;
}

Help.propTypes = {
    text: PropTypes.node.isRequired,
};
