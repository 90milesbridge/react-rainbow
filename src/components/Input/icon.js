import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import IconSvg from '../IconSvg';


export default function Icon(props) {
    const {
        isVisible,
        iconName,
        position,
        error,
    } = props;

    const getIconClassNames = () => {
        if (error) {
            return 'slds-input__icon';
        }
        return classnames(
            'slds-icon slds-input__icon slds-icon-text-default',
            `slds-input__icon_${position}`,
        );
    };

    if (isVisible) {
        return <IconSvg iconName={iconName} className={getIconClassNames()} />;
    }
    return null;
}

Icon.propTypes = {
    iconName: PropTypes.string.isRequired,
    isVisible: PropTypes.bool,
    position: PropTypes.string.isRequired,
    error: PropTypes.node.isRequired,
};

Icon.defaultProps = {
    isVisible: false,
};
