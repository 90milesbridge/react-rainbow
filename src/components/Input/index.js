import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { uniqueId } from './../../libs/utils';
import Icon from './icon';
import Help from './help';
import Error from './error';
import Label from './label';

const inputId = uniqueId('input');
const inlineTextLabelId = uniqueId('inline-text-label');
const errorMessageId = uniqueId('error-message');

export default function Input(props) {
    const {
        className,
        style,
        value,
        onChange,
        label,
        error,
        placeholder,
        disabled,
        readOnly,
        required,
        tabIndex,
        type,
        maxLength,
        minLength,
        pattern,
        labelClassName,
        iconName,
        iconPosition,
        bottomHelpText,
        isBare,
        isCentered,
    } = props;
    const isRequiredOrHasError = !!(required || error);

    const getContainerClassNames = () => classnames('slds-form-element', {
        'slds-has-error': error,
    }, className);

    const getFormControlClassNames = () => classnames('slds-form-element__control', {
        'slds-input-has-icon': iconName,
        [`slds-input-has-icon_${iconPosition}`]: iconName,
    });

    const getInputClassNames = () => classnames('slds-input', {
        'slds-input_bare': isBare,
        'slds-input_counter': isCentered,
    });

    const getInlineTextLabelId = () => {
        if (bottomHelpText) {
            return inlineTextLabelId;
        }
        return undefined;
    };

    const getErrorMessageId = () => {
        if (error) {
            return errorMessageId;
        }
        return undefined;
    };

    return (
        <div className={getContainerClassNames()} style={style}>
            <Label
                className={labelClassName}
                label={label}
                required={isRequiredOrHasError}
                inputId={inputId}
                id={getInlineTextLabelId()} />

            <div className={getFormControlClassNames()}>
                <Icon
                    iconName={iconName}
                    position={iconPosition}
                    error={error} />

                <input
                    id={inputId}
                    type={type}
                    className={getInputClassNames()}
                    value={value}
                    placeholder={placeholder}
                    onChange={onChange}
                    tabIndex={tabIndex}
                    disabled={disabled}
                    readOnly={readOnly}
                    required={isRequiredOrHasError}
                    maxLength={maxLength}
                    minLength={minLength}
                    pattern={pattern}
                    aria-labelledby={getInlineTextLabelId()}
                    aria-describedby={getErrorMessageId()} />

            </div>
            <Help text={bottomHelpText} />
            <Error id={getErrorMessageId()} error={error} />
        </div>
    );
}

Input.propTypes = {
    /** The class name of the root element. */
    className: PropTypes.string,
    /** It is an object with custom style applied to the root element. */
    style: PropTypes.object,
    /** The value of the input */
    value: PropTypes.string,
    /** Event fired when the input value change */
    onChange: PropTypes.func,
    /** The input label */
    label: PropTypes.node,
    /** The error to show below the input. */
    error: PropTypes.node,
    /** The placeholder of the input */
    placeholder: PropTypes.string,
    /** Disables the input if set to true. */
    disabled: PropTypes.bool,
    /** If set to true the input is read only. */
    readOnly: PropTypes.bool,
    /** If set to true the input is required. */
    required: PropTypes.bool,
    /** Tab index */
    tabIndex: PropTypes.number,
    /** The input type */
    type: PropTypes.oneOf([
        'text',
        'password',
        'datetime',
        'datetime-local',
        'date',
        'month',
        'time',
        'week',
        'number',
        'email',
        'url',
        'search',
        'tel',
        'color',
    ]),
    /** The max length of characters allowed by the input. */
    maxLength: PropTypes.number,
    /** The min length of characters allowed by the input. */
    minLength: PropTypes.number,
    /** The input pattern. */
    pattern: PropTypes.string,
    /** The class name of the label element. */
    labelClassName: PropTypes.string,
    /** The name of the icon. Names are written in the
     format '\utility:down\' where 'utility' is the category, and 'down' is the
     specific icon to be displayed. */
    iconName: PropTypes.string,
    /** The position of the icon if it is passed. */
    iconPosition: PropTypes.oneOf([
        'left', 'right',
    ]),
    /** The help text to show below the input. */
    bottomHelpText: PropTypes.node,
    /** If is set to true the input will not have border. */
    isBare: PropTypes.bool,
    /** If is set to true the input text will be centered. */
    isCentered: PropTypes.bool,
};

Input.defaultProps = {
    className: undefined,
    style: {},
    value: undefined,
    onChange: () => {},
    label: null,
    error: null,
    placeholder: null,
    disabled: false,
    readOnly: false,
    required: false,
    tabIndex: undefined,
    type: 'text',
    maxLength: undefined,
    minLength: undefined,
    pattern: undefined,
    labelClassName: undefined,
    iconName: undefined,
    iconPosition: 'left',
    bottomHelpText: null,
    isBare: false,
    isCentered: false,
};
