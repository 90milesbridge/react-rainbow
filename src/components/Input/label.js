import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import RequiredAsterisk from './requiredAsterisk';

export default function Label(props) {
    const {
        className,
        label,
        required,
        inputId,
        id,
    } = props;

    const getLabelClassNames = () => classnames('slds-form-element__label', className);

    if (label) {
        return (
            <label className={getLabelClassNames()} htmlFor={inputId} id={id}>
                <RequiredAsterisk required={required} />
                {label}
            </label>
        );
    }
    return null;
}

Label.propTypes = {
    className: PropTypes.string,
    label: PropTypes.node,
    required: PropTypes.bool.isRequired,
    inputId: PropTypes.string.isRequired,
    id: PropTypes.string,
};

Label.defaultProps = {
    className: undefined,
    label: null,
    id: undefined,
};
