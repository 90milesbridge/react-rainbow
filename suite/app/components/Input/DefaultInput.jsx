import React from 'react';
import { Input } from './../../../../entry'

export default class DefaultInput extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isNumberValid: true
        }
    }
    render() {
        return (
            <div>
                <Input type="text" label="Input label" isValid={ true } />
                <Input type="number" label="Input label" isValid={ this.state.isNumberValid } required={ true } onChange={ e => this.checkValidity(e) } />
                <Input label="Input label" isValid={ true } disabled={ true } />
            </div>
        )
    }

    checkValidity(e) {
        this.setState({ isNumberValid: e.target.checkValidity() })
    }
}