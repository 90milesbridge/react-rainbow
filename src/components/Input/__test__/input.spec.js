import React from 'react';
import { mount } from 'enzyme/build/index';
import renderer from 'react-test-renderer';
import Input from './../index';

describe('<Input/>', () => {
    it('should set the type passed in the Input element', () => {
        const component = mount(
            <Input type="color" />,
        );
        expect(component.find('input').prop('type')).toBe('color');
    });
    it('should set the value passed in the Input element', () => {
        const component = mount(
            <Input value="Input value" />,
        );
        expect(component.find('input').prop('value')).toBe('Input value');
    });
    it('should set the placeholder passed in the Input element', () => {
        const component = mount(
            <Input placeholder="Placeholder Text" />,
        );
        expect(component.find('input').prop('placeholder')).toBe('Placeholder Text');
    });
    it('should set the onChange passed in the Input element', () => {
        const onChangeFn = jest.fn();
        const component = mount(
            <Input onChange={onChangeFn} />,
        );
        expect(component.find('input').prop('onChange')).toBe(onChangeFn);
    });
    it('should set the tabIndex passed in the Input element', () => {
        const component = mount(
            <Input tabIndex={0} />,
        );
        expect(component.find('input').prop('tabIndex')).toBe(0);
    });
    it('should set the disabled passed in the Input element', () => {
        const component = mount(
            <Input disabled />,
        );
        expect(component.find('input').prop('disabled')).toBe(true);
    });
    it('should set the readonly passed in the Input element', () => {
        const component = mount(
            <Input readOnly />,
        );
        expect(component.find('input').prop('readOnly')).toBe(true);
    });
    it('should set the required passed in the Input element', () => {
        const component = mount(
            <Input required />,
        );
        expect(component.find('input').prop('required')).toBe(true);
    });
    it('should set the maxLength passed in the Input element', () => {
        const component = mount(
            <Input maxLength={0} />,
        );
        expect(component.find('input').prop('maxLength')).toBe(0);
    });
    it('should set the minLength passed in the Input element', () => {
        const component = mount(
            <Input minLength={0} />,
        );
        expect(component.find('input').prop('minLength')).toBe(0);
    });
    it('should set the pattern passed in the Input element', () => {
        const component = mount(
            <Input pattern="Input Pattern" />,
        );
        expect(component.find('input').prop('pattern')).toBe('Input Pattern');
    });
    it('should set the bottomHelpText passed in the Help element', () => {
        const component = mount(
            <Input bottomHelpText="Help text" />,
        );
        expect(component.find('Help').prop('text')).toBe('Help text');
    });
    it('should set the error passed in the Error element', () => {
        const component = mount(
            <Input error="Error text" />,
        );
        expect(component.find('Error').prop('error')).toBe('Error text');
    });
    it('should set the error passed in the RequiredAsterisk element', () => {
        const component = mount(
            <Input required />,
        );
        expect(component.find('RequiredAsterisk').prop('required')).toBe(true);
    });
    it('should render an icon on the right when iconName is passed and iconPosition is right', () => {
        const component = mount(
            <Input iconName="utility:activity" iconPosition="right" />,
        );
        expect(component.find('[data-id="input-left-icon"]').prop('isVisible')).toBe(false);
        expect(component.find('[data-id="input-right-icon"]').prop('isVisible')).toBe(true);
    });
    it('should have the right class names when have a custom class', () => {
        const component = renderer.create(
            <Input className="my-custom-class-name" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when iconName is passed', () => {
        const component = renderer.create(
            <Input iconName="utility:activity" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when error is passed', () => {
        const component = renderer.create(
            <Input error="Error text" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when labelClassName is passed', () => {
        const component = renderer.create(
            <Input labelClassName="my-custom-label-class-name" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when isBare', () => {
        const component = renderer.create(
            <Input isBare />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when isCentered', () => {
        const component = renderer.create(
            <Input isCentered />,
        );
        expect(component).toMatchSnapshot();
    });
});
