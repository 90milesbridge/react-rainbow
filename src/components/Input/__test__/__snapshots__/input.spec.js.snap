// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input/> should have the right class names when error is passed 1`] = `
<div
  className="slds-form-element slds-has-error"
  style={Object {}}
>
  <label
    className="slds-form-element__label"
    htmlFor="input-1"
  />
  <div
    className="slds-form-element__control"
  >
    <input
      className="slds-input"
      data-id="input-1"
      disabled={false}
      onChange={[Function]}
      placeholder={null}
      readOnly={false}
      required={false}
      type="text"
    />
  </div>
  <div
    className="slds-form-element__help"
  >
    Error text
  </div>
</div>
`;

exports[`<Input/> should have the right class names when have a custom class 1`] = `
<div
  className="slds-form-element my-custom-class-name"
  style={Object {}}
>
  <label
    className="slds-form-element__label"
    htmlFor="input-1"
  />
  <div
    className="slds-form-element__control"
  >
    <input
      className="slds-input"
      data-id="input-1"
      disabled={false}
      onChange={[Function]}
      placeholder={null}
      readOnly={false}
      required={false}
      type="text"
    />
  </div>
</div>
`;

exports[`<Input/> should have the right class names when iconName is passed 1`] = `
<div
  className="slds-form-element"
  style={Object {}}
>
  <label
    className="slds-form-element__label"
    htmlFor="input-1"
  />
  <div
    className="slds-form-element__control slds-input-has-icon slds-input-has-icon_left"
  >
    <svg
      aria-hidden={true}
      className="slds-icon slds-input__icon slds-icon-text-default slds-input__icon_left"
      style={Object {}}
    >
      <use
        xlinkHref="/icons/utility-sprite/svg/symbols.svg#activity"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
    <input
      className="slds-input"
      data-id="input-1"
      disabled={false}
      onChange={[Function]}
      placeholder={null}
      readOnly={false}
      required={false}
      type="text"
    />
  </div>
</div>
`;

exports[`<Input/> should have the right class names when isBare 1`] = `
<div
  className="slds-form-element"
  style={Object {}}
>
  <label
    className="slds-form-element__label"
    htmlFor="input-1"
  />
  <div
    className="slds-form-element__control"
  >
    <input
      className="slds-input slds-input_bare"
      data-id="input-1"
      disabled={false}
      onChange={[Function]}
      placeholder={null}
      readOnly={false}
      required={false}
      type="text"
    />
  </div>
</div>
`;

exports[`<Input/> should have the right class names when isCentered 1`] = `
<div
  className="slds-form-element"
  style={Object {}}
>
  <label
    className="slds-form-element__label"
    htmlFor="input-1"
  />
  <div
    className="slds-form-element__control"
  >
    <input
      className="slds-input slds-input_counter"
      data-id="input-1"
      disabled={false}
      onChange={[Function]}
      placeholder={null}
      readOnly={false}
      required={false}
      type="text"
    />
  </div>
</div>
`;

exports[`<Input/> should have the right class names when labelClassName is passed 1`] = `
<div
  className="slds-form-element"
  style={Object {}}
>
  <label
    className="slds-form-element__label my-custom-label-class-name"
    htmlFor="input-1"
  />
  <div
    className="slds-form-element__control"
  >
    <input
      className="slds-input"
      data-id="input-1"
      disabled={false}
      onChange={[Function]}
      placeholder={null}
      readOnly={false}
      required={false}
      type="text"
    />
  </div>
</div>
`;
